Show and smooth notch cutout

xabolcs opened the pull request:
While at it, add and `svg` to visualise the change and the cutout itself.

xabolcs committed:
grus: overlay: Show defined cutout and also visualize it as a hidden SVG

Add config_mainBuiltInDisplayCutout to an SVG, which name should start
with a dot to avoid breaking the build.

Just keep them in sync and it will help to visualise the cutout.

Signed-off-by: Szabolcs Hubai <szab.hu@gmail.com>

Rocky7842 commented:
The result is what you posted on xda, right?
In my opinion, users should have as much screen space as possible. Increasing the notch size is not to my liking.

xabolcs commented:
> The result is what you posted on xda, right?

Yes, that was built from this branch.

> Increasing the notch size is not to my liking.

Defining the `config_mainBuiltInDisplayCutoutRectApproximation` is to avoiding this, ain't it?

Here is the lock screen with your 2023-04-22 build. The operator text is `vodafone HU – DIGI HU`:
![Screenshot_20230602-194931_crDroid Home_before](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/5522c212-2f7e-4629-ad73-bb616a735d4d)

And here is the lockscreen with this change
![Screenshot_20230602-200707_crDroid Home_after](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/7f5d49c4-969f-4e2e-8f61-41bb703bd2f6)

And here are two photos:
![IMG_20230602_201201](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/5e6c5ec5-d222-49c6-a9cc-b8073051e586)

And with supermacro ( :see_no_evil: ):
![IMG_20230602_201121](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/940a07ca-84ab-496f-bb2c-cf4139895a8e)

Rocky7842 commented:
But didn't the actual black area get increased? Sorry if I misunderstood something, as I don't have access to my phone to visualize it right now.

xabolcs commented:
I moved the `config_fillMainBuiltInDisplayCutout` change to the first commit now and will build each commit in the weekend.

xabolcs commented:
### Stock 12.0.4 / Android 10
Here is a screenshot from stock 12.0.4 (Android 10) firmware. The status bar is 81px height.
![Screenshot_2023-06-04-09-57-43-306_org mozilla firefox_stock-a10-statusbar-height](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/e75f1c18-196b-4cc2-a944-0090ab6b5472)


### 1st commit: always show defined cutout
Here are pictures from the first commit: enabling `config_fillMainBuiltInDisplayCutout`.
Status bar height is 81px height.
![Screenshot_20230604-090735_Üzenetváltás_01-show-defined-cutout](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/1fb29342-e982-4c35-93df-b26c197f9eef)

And the photos.
![IMG_20230604_090608](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/ad2b30b7-c85b-4b5f-bef5-c8c2c486ab76)
Also with "supermacro":
![IMG_20230604_085714](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/d2eb92af-a5a8-4d91-b50c-88c3c3b6d842)

### 2nd commit: define bounding rectangle
The statusbar height is now 80px.
![Screenshot_20230604-092909_Üzenetváltás_02-cutout-bounding](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/9857f4a2-7fcb-4cb7-b285-73c9062840e5)

Defining the bounding rectangle makes the cutout thinner, so the status bar have more space for its texts and icons.
![IMG_20230604_092131](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/92500244-1b60-4428-b1f3-229eb7be00a5)

Closer look:
![IMG_20230604_092438](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/e70a7e63-7fac-43e2-b14c-26cddf936b92)

### 3rd commit: smoothing the cutout
While keeping the previously defined bounding rectangle modify the `config_mainBuiltInDisplayCutout` config for a better looking cutout.
Both the cutout and the bounding rectangle came from crDroid's **ginkgo** definitions.

Screenshot - no change from previous
![Screenshot_20230604-093833_Üzenetváltás_03-smooth-cutout](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/df92f2ec-8cbb-43f0-b831-bea5390f4b4e)

(Real world pictures are in my [previous comment](https://github.com/Rocky7842/android_device_xiaomi_grus/pull/1#issuecomment-1574110628))

xabolcs commented:
### 12.5.1.0 RFBMIXM / Android 11
Just flashed the 12.5.1.0 RFBMIXM / Android 11 version.
There is a setting about smoothen the notch curves in the stock rom:
![Screenshot_2023-06-04-11-07-41-263_com android settings_adjust-notch-shape](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/4e968b28-7430-4346-bdeb-3bde393f61a2)

![IMG_20230604_111107](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/50eee348-f066-49ea-ac4d-c36b254f7532)

I prefer the smooth one. :slightly_smiling_face: 
![IMG_20230604_111008](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/2dc7796e-a74c-49f2-ac91-ffe1729a2e14)


Btw the status bar is now 80px height (but there could be a blur there, see the Play Music screenshot! :roll_eyes:)
![Screenshot_2023-06-04-11-08-12-424_com android chrome_statusbar](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/405616ee-9ff8-4f13-bf8e-37ee57d922c2)

![Screenshot_2023-06-04-11-09-00-992_com google android music_statusbar](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/3b6ba339-045d-4a03-818d-bff8319b1e7c)

Rocky7842 commented:
The notch we'll have with these commits seems still bigger then smoothed stock notch?

Aside from that, I'm still leaning toward none smoothed notch. So maybe I won't merge this one.
If you can find some way to get us a notch as close as our stock one (non-smoothed) and improve the overall behavior, that would be better.

xabolcs commented:
I think so too.
EDIT: my version is just thicker. It has the same height.


But I don't know how could I generate those curves from raster graphics. Or how else could I test new svg values without building and flashing.

xabolcs commented:
> If you can find some way to get us a notch as close as our stock one (non-smoothed) and improve the overall behavior, that would be better.

Is there a setting in the crDroid tools app, which could do this smooth non-smooth toggle? 🙃

For the record: the smoothed one is the default for that 12.5.1.0 RFBMIXM 😉
But sure, I'll go for the thinner one.

xabolcs commented:
With the help of [Mistkeithy/SetWallpaper](https://github.com/Mistkeithy/SetWallpaper/releases) it wasn't as hard to iterate over and over.

@Rocky7842 , please try the following wallpapers!
* [green notch with transparent / black background](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/a9f7a8ba-f447-48d7-8585-f135d7f83f6f)
* [black notch with white background](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/b40dfa20-1ab5-42f6-babb-3e1eb4338fd3)

----
Meh :roll_eyes: 
They looks good only on stock ROM.

xabolcs commented:
> stuff

Tried it. Its just awesome! :heart_eyes: 
D/L: https://fromsmash.com/iSJaWonJ9h-dt
Expiry: 14th June 2023

xabolcs commented:
And here are a few screenshots from commit dc06c6905f796fb53da22d7d2bd9e1637c1d8c88.

### Narrower stock notch in action
![IMG_20230607_081540](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/d9fcff7f-6ced-41d3-a641-03a67c5a4862)

"Supermicro" Home Screen
![IMG_20230607_081704](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/9c3d42fb-0bc3-41ca-ad07-26116eac99a2)

"Supermicro" Lock Screen
![IMG_20230607_081808](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/a3fedf23-2187-4ed1-8eaa-05604a26b4d3)

Rocky7842 commented:
Not sure but it seems it's still a touch wider than stock? (According to your screenshot.) Still, already looks good to me though.

xabolcs commented:
> Not sure but it seems it's still a touch wider than stock? (According to your screenshot.)

It is, but I didn't want to go more thinner as I wouldn't able to verify how thinner than the stock  my notch actually is.
See real life photos below with the green wallpaper earlier!
My notch is a few pixel wider than the stock. (Transparent background was rendered as black on stock fw.)

![IMG_20230607_111427](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/40d9ed39-2016-4b45-8a22-41ddf113f269)
![IMG_20230607_111410](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/5834330e-ef1f-422a-bebe-8e14468fc812)
![IMG_20230607_111529](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/8faf430c-e003-408e-a275-4a3a5ab86080)
![IMG_20230607_111451](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/c1b02a20-d5f2-441e-a2ef-29c9b409e896)



> Still, already looks good to me though.

:pray:

xabolcs committed:
grus: overlay: Add the cutout bounding rectangle from ginkgo

While at it, visualise and make it to a real rectangle.

See original commit description below:
crdroidandroid/android_device_xiaomi_ginkgo@2dfc3f0ac5d1

ginkgo: overlay: Use modified notch cutout from Mi A3

 * the stock one taken from DevicesOverlay.apk is actually fake and ends up as a rectangle
   cutout which is of no use to us at all, and we even have cutout fill disabled which makes
   it completely pointless; the physical cutout doesn't look that good.

 * Mi A3 has a notch cutout identical to ours so let's use its cutout config which i've
   extracted from framework-res and edited manually using https://yqnn.github.io/svg-path-editor/
   to better match our notch and statusbar height (80px).

 * now our notch cutout looks smooth and anti-aliased without any jagged edges, like on MIUI.

Change-Id: I193428a5656b99639aeacfa5e5bb673b5c05c6ac
Signed-off-by: Szabolcs Hubai <szab.hu@gmail.com>

xabolcs committed:
grus: overlay: Use stock-like thinner notch

Took ginkgo's wider notch cutout to start with. [1]

After then iterate with Yann's SVG editor [2], Inkscape
and Mistkeithy's SetWallpaper app [3] on 12.5.1.0 RFBMIXM (Android 11)
to make the notch thinner enough.

In the stock firmware, there is a toggle with title "Adjust node shape"
and with detail as "Turn on to smoothen notch curves".

Current cutout config goes for the thinner one (smoothen is disabled),
but I kept it a few pixels wider than the stock.

For real world photos take a look at Rocky7842's tree! [4]

[1]: crdroidandroid/android_device_xiaomi_ginkgo@c534f7685b7b
[2]: https://yqnn.github.io/svg-path-editor/
[3]: https://github.com/Mistkeithy/SetWallpaper
[4]: Rocky7842/android_device_xiaomi_grus#1

Signed-off-by: Szabolcs Hubai <szab.hu@gmail.com>

xabolcs committed:
grus: overlay: Status bar is 80px height since Android 11

While the status bar height was 81px in 12.0.4.0 QFBMIXM - Android 10,
now it is 80px in 12.5.1.0 RFBMIXM - Android 11.

Anyway, this is not really used since the cutout rectangle is defined:
config_mainBuiltInDisplayCutoutRectApproximation

Signed-off-by: Szabolcs Hubai <szab.hu@gmail.com>

xabolcs commented:
@Rocky7842 , <del>this is ready now, please test and review! Thanks!</del>

xabolcs commented:
> Not sure but it seems it's still a touch wider than stock?

I was able to use correctly my test wallpapers on crDroid too, so I'll try a few more rounds to make it more close to the stock.
Going to shrink the bounding rectangle too.

xabolcs commented:
> I'll try a few more rounds to make it more close to the stock.

Tried it. Doesn't look good: the curves were too angular. :confused: 
> Going to shrink the bounding rectangle too.

Shrinked the rectangle's width from 132px to 120px.

Rocky7842 commented:
I'm still thinking about how to properly review this, considering that one of my relatives now uses the phone as his daily driver.
Making matters worse, the screen protector on it has a notch bigger than it should, and I don't have a spare one to remove and reapply it.

xabolcs commented:
Should I create non-listed Youtube videos and put the links here?
If you write testing steps, I could do them while recording the screen with screenrecord and with a real camera.

xabolcs commented:
If you allow me, I could post a build of this PR to your XDA forum, to ask brave users for test. 🙂

Rocky7842 commented:
> If you allow me, I could post a build of this PR to your XDA forum, to ask brave users for test. 🙂

Good idea. I've posted the comment. Will tell you when someone pm me.

xabolcs commented:
@Rocky7842, I'm having fun with creating simple Android apps which lists themself in the cutout menu of the developer tools.
Of course my apk ain't work: doesn't appear in the list. :roll_eyes: 
If you are interested, please have a [look](https://github.com/xabolcs/GrusCutoutOverlay)!

It would be nice if it would work, as it would allow testing without flashing a full update: just flash it through recovery or install with Magisk.

Rocky7842 commented:
@xabolcs 
`<string translatable="false" name="config_mainBuiltInDisplayCutout">
M -81,0 L -39,81 L 39,81 L 81,0 Z
</string>`
Can you try this for me?

xabolcs commented:
Sorry for the confusion: here or in the linked Apk repository?

Anyway, these are straight lines, to the height of 81px which is deprecated ~~of~~ since Android 11 based stock MIUI.


(Unable to attach pic, currently)

Rocky7842 commented:
> Sorry for the confusion: here or in the linked Apk repository?
> 
> Anyway, these are straight lines, to the height of 81px which is deprecated of stock Android 11.
> 
> 
> (Unable to attach pic, currently)

I'm just wondering how this looks. This is from official lineage.
So apparently it is what I imagine it would look like. They probably don't want to take the effort to match the real notch.

xabolcs commented:
Here you are!

![image](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/7f6e58df-c728-4cbd-9294-2177c159a15d)

Will do a full build tonight (8 hrs from now).


> They probably don't want to take the effort to match the real notch.

Yup, I think too.

xabolcs commented:
> They probably don't want to take the effort to match the real notch.

In the end, I would like to upstream my notches to LineageOS. But before it, it would be nice to test it widely.

xabolcs commented:
@Rocky7842 

> I'm just wondering how this looks. This is from official lineage.

[Here](https://github.com/Rocky7842/android_device_xiaomi_grus/compare/crdroid-13.0...xabolcs:android_device_xiaomi_grus:notch-cutout-lineageos) is the diff!

> `<string translatable="false" name="config_mainBuiltInDisplayCutout"> M -81,0 L -39,81 L 39,81 L 81,0 Z </string>`

Ouch! :see_no_evil:
![IMG_20230626_090612](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/8d0ab1ee-ed07-4f2d-8428-1ab9aa91a1da)

xabolcs commented:
@Rocky7842 
Just finished the zip installable version: [GrusCutoutOverlay](https://github.com/xabolcs/GrusCutoutOverlay/releases) :tada: 

Just install with Magisk (or recovery) and check the Display cutout menu in the Developer Tools! :ok_hand:

Rocky7842 committed:
grus: Add smooth cutout option under developer options

Szabolcs improved the smoothened version of the cutout [1] with the new
DisplayCutoutViewer [2], let's include it.

[1]: https://github.com/xabolcs/GrusCutoutOverlay/releases/tag/v2.0
[2]: https://github.com/SebaUbuntu/DisplayCutoutViewer

Co-developed-by: Szabolcs Hubai <szab.hu@gmail.com>

Rocky7842 commented:
@xabolcs You might want to take a look at my `crdroid-13.0-cutout` branch. I've taken the value from `dfff88f8ee60428eb92f35c2d0ecb870f031ca3c` since I haven't been able to validate it yet. My plan is to use the stock thin as the default and utilize this method to allow for the use of a smoothed one.

xabolcs commented:
Yeah thats was I tought too: put these paths into Developer Options for a few monts, without changing the default - let the users test them! 😁
I just didn't have time to do that PR.
I wanted to do the same for **elpaablo**'s AlphaDroid.

And in the end, I wanted to change the default in LineageOS (through gerrit).


___

I didn't tested your changes, just looked at it, but I would go without changing the default, but sure, it's your fork - your choice! :+1:

xabolcs commented:
And yeah, the smooth one is came from `willow`/`ginkgo` which looks good on `willow` but looks too wide o  `grus`. :confused:
I had to do a few iteration with it to have a better shape.

xabolcs commented:
Hi @Rocky7842 !

Rebased and included your `rro_overlays` commit to add the smoothened one too! Thank for the great idea! :ok_hand: 
I incorporated my changes from my [GrusCutoutOverlay](https://github.com/xabolcs/GrusCutoutOverlay/releases/tag/v2.0) so now both notch version look awesome! :star_struck: 

Half-offtopic question: may I add hungarian translation for this "Smooth" cutout?

xabolcs commented:
Just to note: not tried yet ... I don't even compiled it, just rebased.

Rocky7842 commented:
> Hi @Rocky7842 !
> 
> Rebased and included your `rro_overlays` commit to add the smoothened one too! Thank for the great idea! 👌 I incorporated my changes from my [GrusCutoutOverlay](https://github.com/xabolcs/GrusCutoutOverlay/releases/tag/v2.0) so now both notch version look awesome! 🤩
> 
> Half-offtopic question: may I add hungarian translation for this "Smooth" cutout?

Any picture of the new ones?

xabolcs commented:
The "default" one didn't change:
![IMG_20230826_071223-01](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/6ba68d55-17a9-4bb7-9191-e1d926b3f65e)

I improved only the smoothened one:
![IMG_20230826_071312-01](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/08a78050-bbbc-47a6-b5f5-841b1ba758fd)

Both can be tested as Magisk / KernelSU module from my another repo.

Rocky7842 commented:
Looks good.
I will however tune the default notch to shrink it a little bit, as I found it is unnecessarily big in stock ROMs. I prefer it to be just a touch bigger than the black zone of the top glass. Similar to [this](https://github.com/Rocky7842/android_device_xiaomi_xmsirius/commit/285e8d85791f7e01421a9029f3e95fce243e119d). So the smoothened one may or may not require a little tuning on the bottom side too.
Anyways, thank you for your effort and these notches look real good. I'll tune it like mentioned above and merge it in the future once the phone is available.

Rocky7842 commented:
It will stay like https://github.com/Rocky7842/android_device_xiaomi_grus/commit/983ffc8676dd88954074c071550c54c3ba20dabf before I tune the default one.

xabolcs commented:
Sure!
Let me shoot a few photos with stock rom to show how few pixels were left between the stok notches and my paths. Sure, will do the same with crDroid too.
Let me fix the co-stuff tag too.

xabolcs commented:
Here is my stock-like path on crDroid.
![IMG_20230828_082652-01](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/f3d84ca4-de47-4397-b161-e11889c6fadf)

xabolcs commented:
And here is a video about it:

https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/45584e53-934a-45e6-b408-0418ab188917

xabolcs commented:
Here are my paths on stock MIUI Global 12.5.1.0 (RFBEUXM).

"stock":
![IMG_20230830_174233-01](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/4b1dc597-cb67-4e8e-8701-920ed66ee559)


"smoothened stock":
![IMG_20230830_174439-01](https://github.com/Rocky7842/android_device_xiaomi_grus/assets/568751/fc46c443-c070-4bb6-a2b1-3052a6d8dc9f)

> I will however tune the default notch to shrink it a little bit, as I found it is unnecessarily big in stock ROMs. I prefer it to be just a touch bigger than the black zone of the top glass.

I'll try to shrink the stock one a little bit and create a new one, "top glass"
